source/common/ur_pool_manager.hpp

```diff
@@ -158,63 +194,83 @@ inline bool pool_descriptor::operator==(const pool_descriptor &other) const {
            lhs.poolHandle == rhs.poolHandle;
 }
 
 inline std::ostream &operator<<(std::ostream &os, const pool_descriptor &desc) {
     os << "pool handle: " << desc.poolHandle
        << " context handle: " << desc.hContext
        << " device handle: " << desc.hDevice << " memory type: " << desc.type
-       << " is read only: " << desc.deviceReadOnly;
+       << " is read only: "
+       << (desc.deviceFlags & UR_USM_DEVICE_MEM_FLAG_DEVICE_READ_ONLY);
     return os;
 }
 
 inline std::pair<ur_result_t, std::vector<pool_descriptor>>
-pool_descriptor::create(ur_usm_pool_handle_t poolHandle,
-                        ur_context_handle_t hContext) {
+pool_descriptor::createDefaults(ur_usm_pool_handle_t poolHandle,
+                                ur_context_handle_t hContext) {
     auto [ret, devices] = urGetAllDevicesAndSubDevices(hContext);
     if (ret != UR_RESULT_SUCCESS) {
         return {ret, {}};
     }
 
     std::vector<pool_descriptor> descriptors;
-    pool_descriptor &desc = descriptors.emplace_back();
-    desc.poolHandle = poolHandle;
-    desc.hContext = hContext;
-    desc.type = UR_USM_TYPE_HOST;
+    descriptors.emplace_back(poolHandle, hContext, nullptr, UR_USM_TYPE_HOST,
+                             nullptr);
 
     for (auto &device : devices) {
         {
-            pool_descriptor &desc = descriptors.emplace_back();
-            desc.poolHandle = poolHandle;
-            desc.hContext = hContext;
-            desc.hDevice = device;
-            desc.type = UR_USM_TYPE_DEVICE;
+            descriptors.emplace_back(poolHandle, hContext, device,
+                                     UR_USM_TYPE_DEVICE, nullptr);
         }
         {
-            pool_descriptor &desc = descriptors.emplace_back();
-            desc.poolHandle = poolHandle;
-            desc.hContext = hContext;
-            desc.type = UR_USM_TYPE_SHARED;
-            desc.hDevice = device;
-            desc.deviceReadOnly = false;
+            descriptors.emplace_back(poolHandle, hContext, device,
+                                     UR_USM_TYPE_SHARED, nullptr);
         }
         {
-            pool_descriptor &desc = descriptors.emplace_back();
-            desc.poolHandle = poolHandle;
-            desc.hContext = hContext;
-            desc.type = UR_USM_TYPE_SHARED;
-            desc.hDevice = device;
-            desc.deviceReadOnly = true;
+            ur_usm_device_desc_t deviceDesc = {
+                UR_STRUCTURE_TYPE_USM_DEVICE_DESC, nullptr,
+                UR_USM_DEVICE_MEM_FLAG_DEVICE_READ_ONLY};
+            ur_usm_desc_t desc = {UR_STRUCTURE_TYPE_USM_DESC, &deviceDesc, 0,
+                                  0};
+
+            descriptors.emplace_back(poolHandle, hContext, device,
+                                     UR_USM_TYPE_SHARED, &desc);
         }
     }
 
     return {ret, descriptors};
 }
 
+inline usm::DisjointPoolMemType
+urTypeToDisjointPoolType(const ur_usm_type_t &type) {
```

Comment:
What about SharedReadOnly?

Reply:
I think only having these three types makes sense because they correspond to the three allocation entry points in UR, read only pools will still be distinct by the fact that they're TYPE_SHARED and pass the appropriate read only alloc flag

Reply:
The problem is that Shared and SharedReadOnly pools have different defaults (specifically, for Shared pooling is disabled by [default](https://github.com/oneapi-src/unified-runtime/blob/main/source/common/umf_pools/disjoint_pool_config_parser.cpp#L53)) and we also allow users to change the pooling behavior for each of those 4 types separately through UR_L0_USM_ALLOCATOR (aka SYCL_PI_LEVEL_ZERO_USM_ALLOCATOR). This env variable is not very well documented and is still considered 'experimental' but it's been useful for some applications.

Reply:
Oh I see.. I was worried that the extra type was more than it first seemed. This aspect probably needs a re-design in that case, I'll be revisiting this PR once the pool manager PRs merge and I can rebase